### PR TITLE
Migrate to Swift 3 [WIP]

### PR DESCRIPTION
**Pull request by limfinity**
## Information

This is a WIP merge to track the difference between develop and Swift 3.

## How to test

Checkout project and run example app. There should be no differences in behaviour and UI compared to other Swift versions.

## Merge information

Should *not* be merged until future handling of Swift versions is determined. 

## YouTrack Tickets

MOBILE-657

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -8,20 +8,21 @@ env:
...
+    - osx_image: xcode8
+      env: DESTINATION="OS=9.3,name=iPhone 6 Plus"      SCHEME="$IOS_FRAMEWORK_SCHEME"     SDK="$IOS_SDK"     RUN_TESTS="YES"     POD_LINT="NO"      BUILD_DOCS="NO"      BUILD_COVERAGE="NO"
+    - osx_image: xcode8
+      env: DESTINATION="OS=10.0,name=iPhone 6S"         SCHEME="$IOS_FRAMEWORK_SCHEME"     SDK="$IOS_SDK"     RUN_TESTS="YES"     POD_LINT="NO"      BUILD_DOCS="NO"      BUILD_COVERAGE="YES"
 before_install:
   - gem install slather
-  - gem install cocoapods # Since Travis is not always on latest version
+   # Pre-Release version of Cocoapods needed for Xcode8 environment.
+   # Can be removed when v1.1.0 is officially released.
+  - gem install cocoapods --pre # Since Travis is not always on latest version
```
Comment (limfinity) [inline .travis.yml]: We can now remove `-- pre` and the comments as Cocoapods has been updated to version 1.1.1.

File: Example/GiniVision/AnalysisManager.swift [hunk 2]
```diff
@@ -135,51 +135,51 @@ class AnalysisManager {
...
                 self.error = error
                 userInfo = [ GINIAnalysisManagerErrorUserInfoKey: error ]
                 notificationName = GINIAnalysisManagerDidReceiveErrorNotification
-                completion?(inner: { _ in throw AnalysisError.Unknown })
+                completion?({ _ in throw AnalysisError.unknown })
                 return nil
             }
-            dispatch_async(dispatch_get_main_queue(), {
-                NSNotificationCenter.defaultCenter().postNotificationName(notificationName, object: self, userInfo: userInfo)
+            DispatchQueue.main.async(execute: {
```
Comment (limfinity) [inline Example/GiniVision/AnalysisManager.swift]: Let's be consistent throughout the app and framework and omit the braces `(execute: ... )`.

File: Example/GiniVision/ComponentAPIAnalysisViewController.swift [hunk 3]
```diff
@@ -164,18 +164,18 @@ class ComponentAPIAnalysisViewController: UIViewController {
         guard errorButton.alpha != 1.0 else {
             return
         }
-        UIView.animateWithDuration(0.5) {
+        UIView.animate(withDuration: 0.5, animations: {
```
Comment (limfinity) [inline Example/GiniVision/ComponentAPIAnalysisViewController.swift]: Here we could also drop the `animations:` keyword and let the closure stand alone at the end. 
What style would you embrace? @nsobadjiev

File: Example/GiniVision/ComponentAPIAnalysisViewController.swift [hunk 4]
```diff
@@ -164,18 +164,18 @@ class ComponentAPIAnalysisViewController: UIViewController {
...
-        }
+        }) 
     }
     
     func hideErrorButton() {
         guard errorButton.alpha != 0.0 else {
             return
         }
-        UIView.animateWithDuration(0.5) { 
+        UIView.animate(withDuration: 0.5, animations: {
```
Comment (limfinity) [inline Example/GiniVision/ComponentAPIAnalysisViewController.swift]: See above.

File: Example/GiniVision/ComponentAPICameraViewController.swift [hunk 5]
```diff
@@ -36,8 +36,8 @@ class ComponentAPICameraViewController: UIViewController {
         contentController = GINICameraViewController(success:
             { imageData in
                 self.imageData = imageData
-                dispatch_async(dispatch_get_main_queue(), { 
-                    self.performSegueWithIdentifier("showReview", sender: self)
+                DispatchQueue.main.async(execute: {
```
Comment (limfinity) [inline Example/GiniVision/ComponentAPICameraViewController.swift]: See above.

File: Example/GiniVision/ScreenAPIViewController.swift [hunk 6]
```diff
@@ -84,13 +84,13 @@ class ScreenAPIViewController: UIViewController {
         let vc = GINIVision.viewController(withDelegate: self, withConfiguration: giniConfiguration)
         
         // 3. Present the Gini Vision Library Screen API modally
-        presentViewController(vc, animated: true, completion: nil)
+        self.present(vc, animated: true, completion: nil)
```
Comment (limfinity) [inline Example/GiniVision/ScreenAPIViewController.swift]: Omit `self` if not needed.

File: GiniVision/Classes/GINIBarButtonItem.swift [hunk 7]
```diff
@@ -26,13 +26,13 @@ internal class GINIBarButtonItem: UIBarButtonItem {
         // Set accessibility label on all elements
         self.accessibilityLabel = title
         
-        var attributes = titleTextAttributesForState(.Normal) ?? [String : AnyObject]()
+        var attributes = titleTextAttributes(for: UIControlState()) ?? [String : AnyObject]()
         attributes[NSFontAttributeName] = GINIConfiguration.sharedConfiguration.navigationBarItemFont
-        setTitleTextAttributes(attributes, forState: .Normal)
+        setTitleTextAttributes(attributes, for: UIControlState())
```
Comment (limfinity) [inline GiniVision/Classes/GINIBarButtonItem.swift]: I'd rather use `.normal` although they result in the same value for now.

File: GiniVision/Classes/GINICamera.swift [hunk 8]
```diff
@@ -16,80 +16,80 @@ internal class GINICamera {
...
+    fileprivate lazy var motionManager = GINIMotionManager()
     
     init() throws {
         try setupSession()
     }
     
     // MARK: Public methods
     func start() {
-        dispatch_async(sessionQueue, {
+        sessionQueue.async(execute: {
```
Comment (limfinity) [inline GiniVision/Classes/GINICamera.swift]: See above.

File: GiniVision/Classes/GINICamera.swift [hunk 9]
```diff
@@ -16,80 +16,80 @@ internal class GINICamera {
...
             self.session.startRunning()
             self.motionManager.startDetection()
         })
     }
     
     func stop() {
-        dispatch_async(sessionQueue, {
+        sessionQueue.async(execute: {
```
Comment (limfinity) [inline GiniVision/Classes/GINICamera.swift]: See above.

File: GiniVision/Classes/GINICamera.swift [hunk 10]
```diff
@@ -16,80 +16,80 @@ internal class GINICamera {
...
             self.session.stopRunning()
             self.motionManager.stopDetection()
         })
     }
     
-    func focusWithMode(focusMode: AVCaptureFocusMode, exposeWithMode exposureMode: AVCaptureExposureMode, atDevicePoint point: CGPoint, monitorSubjectAreaChange: Bool) {
-        dispatch_async(sessionQueue, {
+    func focusWithMode(_ focusMode: AVCaptureFocusMode, exposeWithMode exposureMode: AVCaptureExposureMode, atDevicePoint point: CGPoint, monitorSubjectAreaChange: Bool) {
+        sessionQueue.async(execute: {
```
Comment (limfinity) [inline GiniVision/Classes/GINICamera.swift]: See above.

File: GiniVision/Classes/GINICamera.swift [hunk 11]
```diff
@@ -16,80 +16,80 @@ internal class GINICamera {
...
-            device.subjectAreaChangeMonitoringEnabled = monitorSubjectAreaChange
+            device.isSubjectAreaChangeMonitoringEnabled = monitorSubjectAreaChange
             device.unlockForConfiguration()
         })
     }
     
-    func captureStillImage(completion: (inner: () throws -> NSData) -> ()) {
-        dispatch_async(sessionQueue, {
+    func captureStillImage(_ completion: @escaping (_ inner: () throws -> Data) -> ()) {
+        sessionQueue.async(execute: {
```
Comment (limfinity) [inline GiniVision/Classes/GINICamera.swift]: See above.

File: GiniVision/Classes/GINICamera.swift [hunk 12]
```diff
@@ -16,80 +16,80 @@ internal class GINICamera {
...
             }
-            self.videoDeviceInput?.device.setFlashModeSecurely(.On)
-            self.stillImageOutput?.captureStillImageAsynchronouslyFromConnection(connection, completionHandler: { (imageDataSampleBuffer: CMSampleBuffer!, error: NSError!) -> Void in
-                guard error == nil else { return completion(inner: { _ in throw GINICameraError.CaptureFailed }) }
+            self.videoDeviceInput?.device.setFlashModeSecurely(.on)
+            self.stillImageOutput?.captureStillImageAsynchronously(from: connection, completionHandler: { (imageDataSampleBuffer: CMSampleBuffer?, error: Error?) -> Void in
+                guard error == nil else { return completion({ _ in throw GINICameraError.captureFailed }) }
                 let imageData = AVCaptureStillImageOutput.jpegStillImageNSDataRepresentation(imageDataSampleBuffer)
-                completion(inner: { _ in return imageData })
+                completion({ _ in return imageData! })
```
Comment (limfinity) [inline GiniVision/Classes/GINICamera.swift]: Not sure what kind of problems could arise here, but I'm not a fan of force unwrapping here.

File: GiniVision/Classes/GINICamera.swift [hunk 13]
```diff
@@ -99,23 +99,23 @@ internal class GINICamera {
     }
     
     // MARK: Private methods
-    private func setupSession() throws {
+    fileprivate func setupSession() throws {
         // Setup is not performed asynchronously because of KVOs
-        func deviceWithMediaType(mediaType: String, preferringPosition position: AVCaptureDevicePosition) -> AVCaptureDevice? {
-            let devices = AVCaptureDevice.devicesWithMediaType(mediaType).filter { $0.position == position }
+        func deviceWithMediaType(_ mediaType: String, preferringPosition position: AVCaptureDevicePosition) -> AVCaptureDevice? {
+            let devices = AVCaptureDevice.devices(withMediaType: mediaType).filter { ($0 as AnyObject).position == position }
```
Comment (limfinity) [inline GiniVision/Classes/GINICamera.swift]: We should cast to `AVCaptureDevice` to be more explicit.

File: GiniVision/Classes/GINICameraContainerViewController.swift [hunk 14]
```diff
@@ -39,12 +39,12 @@ internal class GINICameraContainerViewController: UIViewController, GINIContaine
                 delegate?.didCapture(imageData)
                 
                 // Push review container view controller
-                dispatch_async(dispatch_get_main_queue(), { () -> Void in
+                DispatchQueue.main.async(execute: { () -> Void in
```
Comment (limfinity) [inline GiniVision/Classes/GINICameraContainerViewController.swift]: See above.

File: GiniVision/Classes/GINICameraNotAuthorizedView.swift [hunk 15]
```diff
@@ -11,36 +11,36 @@ import UIKit
...
         label.textColor = GINIConfiguration.sharedConfiguration.cameraNotAuthorizedTextColor
-        label.textAlignment = .Center
+        label.textAlignment = .center
         label.font = GINIConfiguration.sharedConfiguration.cameraNotAuthorizedTextFont
         
         // Configure button
-        button.setTitle(GINIConfiguration.sharedConfiguration.cameraNotAuthorizedButtonTitle, forState: .Normal)
-        button.setTitleColor(GINIConfiguration.sharedConfiguration.cameraNotAuthorizedButtonTitleColor, forState: .Normal)
-        button.setTitleColor(GINIConfiguration.sharedConfiguration.cameraNotAuthorizedButtonTitleColor.colorWithAlphaComponent(0.8), forState: .Highlighted)
+        button.setTitle(GINIConfiguration.sharedConfiguration.cameraNotAuthorizedButtonTitle, for: UIControlState())
```
Comment (limfinity) [inline GiniVision/Classes/GINICameraNotAuthorizedView.swift]: See above.

File: GiniVision/Classes/GINICameraNotAuthorizedView.swift [hunk 16]
```diff
@@ -11,36 +11,36 @@ import UIKit
...
+        button.setTitleColor(GINIConfiguration.sharedConfiguration.cameraNotAuthorizedButtonTitleColor, for: UIControlState())
```
Comment (limfinity) [inline GiniVision/Classes/GINICameraNotAuthorizedView.swift]: See above.

File: GiniVision/Classes/GINIConfiguration.swift [hunk 17]
```diff
@@ -65,7 +65,7 @@ import UIKit
      
      - note: Screen API only.
      */
-    public var navigationBarItemFont = UIBarButtonItem.appearance().titleTextAttributesForState(.Normal)?[NSFontAttributeName] as? UIFont ?? UIFontPreferred(.Regular, andSize: 16)
+    public var navigationBarItemFont = UIBarButtonItem.appearance().titleTextAttributes(for: UIControlState())?[NSFontAttributeName] as? UIFont ?? UIFontPreferred(.regular, andSize: 16)
```
Comment (limfinity) [inline GiniVision/Classes/GINIConfiguration.swift]: See above.

File: GiniVision/Classes/GINIMetaInformationManager.swift [hunk 18]
```diff
@@ -94,7 +94,7 @@ internal extension NSMutableDictionary {
         let keys = stringKeys
         guard keys.count > 0 else { return nil }
         if keys.contains(key) {
-            return self[key]
+            return self.object(forKey: key) as AnyObject?
```
Comment (limfinity) [inline GiniVision/Classes/GINIMetaInformationManager.swift]: We could still use the subscript method.

File: GiniVision/Classes/GINIMotionManager.swift [hunk 19]
```diff
@@ -66,9 +66,9 @@ internal class GINIMotionManager {
         
         currentOrientation = orientationNew
         
-        dispatch_async(dispatch_get_main_queue(), {
-            NSNotificationCenter.defaultCenter().postNotificationName(self.MotionOrientationChangedNotification, object: nil, userInfo: [self.MotionOrientationKey: self])
+        DispatchQueue.main.async(execute: {
```
Comment (limfinity) [inline GiniVision/Classes/GINIMotionManager.swift]: See above.

File: GiniVision/Classes/GININoticeView.swift [hunk 20]
```diff
@@ -76,36 +76,36 @@ internal class GININoticeView: UIView {
...
             }
         }
     }
     
     // MARK: Toggle options
-    func show(animated: Bool = true) {
+    func show(_ animated: Bool = true) {
         if animated {
-            UIView.animateWithDuration(0.5, animations: {
+            UIView.animate(withDuration: 0.5, animations: {
```
Comment (limfinity) [inline GiniVision/Classes/GININoticeView.swift]: See above.

File: GiniVision/Classes/GINIReviewViewController.swift [hunk 21]
```diff
@@ -98,25 +98,25 @@ public typealias GINIReviewErrorBlock = (error: GINIReviewError) -> ()
...
         topView = GININoticeView(text: GINIConfiguration.sharedConfiguration.reviewTextTop)
         
         // Configure bottom view
-        bottomView.backgroundColor = GINIConfiguration.sharedConfiguration.reviewBottomViewBackgroundColor.colorWithAlphaComponent(0.8)
+        bottomView.backgroundColor = GINIConfiguration.sharedConfiguration.reviewBottomViewBackgroundColor.withAlphaComponent(0.8)
         
         // Configure rotate button
-        rotateButton.setImage(rotateButtonImage, forState: .Normal)
-        rotateButton.addTarget(self, action: #selector(rotate), forControlEvents: .TouchUpInside)
+        rotateButton.setImage(rotateButtonImage, for: UIControlState())
```
Comment (limfinity) [inline GiniVision/Classes/GINIReviewViewController.swift]: See above.

File: README.md [hunk 22]
```diff
@@ -78,6 +78,10 @@ GINIVision.setConfiguration(giniConfiguration)
 
 We are providing example apps for Swift and Objective-C. These apps demonstrate how to integrate the Gini Vision Library with the Screen API and Component API. The Gini API SDK is used to analyze the photos of documents. To run the example projects, clone the repo and run `pod install` from the Example or ExampleObjC directory first.
 
+## Swift versions
+
+The Gini Vision Library is entirely written in Swift 2 and needs to be compiled with Xcode versions 7.X. However, you can find both Swift 2.3 and Swift 3 support in different branches within the repository. Please refer to the [Branches](#Branches) section later in this document.
```
Comment (limfinity) [inline README.md]: Add a note on how to use a Swift branch as seen in the branch `swift-2.3` already.

File: README.md [hunk 23]
```diff
@@ -123,6 +127,24 @@ Use this header in an Objective-C project by adding
...
+
+### Develop
+
+The bleeding edge branch. Contains features actively in development and might be unstable
+
+### Swift-2.3
+
+Swift 2.3 support.
+
+### Swift 3
```
Comment (limfinity) [inline README.md]: We should maybe redo the naming to Swift-3.0?

File: README.md [hunk 24]
```diff
@@ -123,6 +127,24 @@ Use this header in an Objective-C project by adding
 
 to your implementation or header files. Note that spaces in your project name result in underscores. So `Your Project` becomes `Your_Project-Swift.h`.
 
+## Branches
```
Comment (limfinity) [inline README.md]: A list would be sufficient or we add more description. Right now it's too much heading for not enough text in my opinion.

**Comment by limfinity** [comment]
@nsobadjiev I renamed the branch could you still look at the issues discussed? Also I'll open a new PR for the new branch.

File: Example/GiniVision/AnalysisManager.swift (hunk 2 above)
Comment (nsobadjiev) [inline Example/GiniVision/AnalysisManager.swift]: Completely agree. I'll fix this and all other cases

File: Example/GiniVision/ScreenAPIViewController.swift (hunk 6 above)
Comment (nsobadjiev) [inline Example/GiniVision/ScreenAPIViewController.swift]: Maybe it's just an Objective-C habit, but I prefer seeing the `self.` so I know it's a member of the same class. Anyway, I've omitted it in this case.

File: GiniVision/Classes/GINIBarButtonItem.swift (hunk 7 above)
Comment (nsobadjiev) [inline GiniVision/Classes/GINIBarButtonItem.swift]: I think the Migration tool made that. I'll change it to `.normal` ofc

File: Example/GiniVision/ScreenAPIViewController.swift (hunk 6 above)
Comment (limfinity) [inline Example/GiniVision/ScreenAPIViewController.swift]: I feel you, but it is part of the current code style I would like to encourage.

File: README.md (hunk 23 above)
Comment (limfinity) [inline README.md]: This is obsolete, as we already communicated according links to clients. We'll tackle this again, when moving Swift 3 to a more prominent place in the repository.

File: README.md [hunk 25]
```diff
@@ -81,12 +80,24 @@ We are providing example apps for Swift and Objective-C. These apps demonstrate
...
 
 ## Installation
 
 Gini Vision Library can either be installed by using CocoaPods or by manually dragging the required files to your project.
 
+**Note**: Irrespective of the option you choose if you want to support **iOS 10** you need to specify the `NSCameraUsageDescription` key in your `Info.plist` file. This key is mandatory for all apps since iOS 10 when using the `Camera` framework. Also if you're using the [Gini iOS SDK](https://github.com/gini/gini-sdk-ios) you need to add support for "Keychain Sharing" in your entitlements by adding a `keychain-access-groups` value to your entitlements file. For more information see the [Integration Guide](http://developer.gini.net/gini-sdk-ios/docs/guides/getting-started.html#integrating-the-gini-sdk) of the Gini iOS SDK.
+
+### Swift versions
+
+The Gini Vision Library is entirely written in **Swift 2.2** and needs to be compiled with Xcode versions 7.X. However, you can find both **Swift 2.3** and **Swift 3** support in different branches within the repository. Please refer to the [Branches](#branches) section later in this document.
```
Comment (a-szotyori) [inline README.md]: If we want to merge this into `develop` we should update this description. With merging into `develop` we prepare for making Swift 3 the default language and Swift 2.2 is not supported anymore (right?) and only Swift 2.3 will stay in its branch.

File: README.md [hunk 26]
```diff
@@ -81,12 +80,24 @@ We are providing example apps for Swift and Objective-C. These apps demonstrate
...
+
+If you use CocoaPods you can specify a branch with:
+
+```ruby
+pod 'GiniVision', :git => 'https://github.com/gini/gini-vision-lib-ios.git', :branch => 'swift-2.3' # or use 'swift3'
```
Comment (a-szotyori) [inline README.md]: Here, too the swift3 branch comment can be removed.

File: README.md [hunk 27]
```diff
@@ -123,6 +134,13 @@ Use this header in an Objective-C project by adding
 
 to your implementation or header files. Note that spaces in your project name result in underscores. So `Your Project` becomes `Your_Project-Swift.h`.
 
+## Branches
+
+* `master` - Contains the latest stable release of the vision library.
```
Comment (a-szotyori) [inline README.md]: This also needs to be updated.

**Pull request closed without merging**